### PR TITLE
Ensure all toolchain requirements are used as task inputs

### LINKED ISSUE #13970: Toolchain is taken into account for up-to-date checks
Currently, toolchains should be resolved based on other inputs to a task (e.g. language level, source compatibility). We need to consider whether the toolchain itself (or properties thereof) should be considered as task inputs as well. For example: is it enough to have the java version as @Input and ignore the toolchain (even though different users may resolve to a different vendor). 
Currently, task output caching is disabled altogether if you use things like `compileTask.options.forkOptions.executable`

### PR DESCRIPTION
**Pull request by bmuskalla**
Fixes #13970

As discussed, we now consider vendor and implementation as well as task inputs but only if they've been specified as toolchain requirements.

### CODE REVIEW

**Comment by bmuskalla** [comment]
@ljacomet Would be good to get this reviews but I'll change its base to `release` once we branched

**Comment by bmuskalla** [comment]
@bot-gradle test this

File: subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java [hunk 1]
```diff
@@ -89,7 +89,7 @@ private JavaToolchain downloadToolchain(JavaToolchainSpec spec) {
         return () -> new GradleException("Provisioned toolchain '" + javaHome.get() + "' could not be probed.");
     }
 
-    private Optional<JavaToolchain> asToolchain(File javaHome) {
-        return toolchainFactory.newInstance(javaHome);
+    private Optional<JavaToolchain> asToolchain(File javaHome, JavaToolchainSpec spec) {
+        return toolchainFactory.newInstance(javaHome, spec);
```
Comment (ljacomet) [inline subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java:93]: Maybe we should convert the spec to the input type here instead.

**Review by ljacomet — APPROVED** [review]
Looks good to me, I like the approach.

One potential suggestion:

**Comment by bmuskalla** [comment]
@bot-gradle test this

**Pull request merged**